### PR TITLE
[Coral-773] Revise data skew policy

### PR DESCRIPTION
**Pull request by jeongyooneo**
This PR:
- Sets hash range as the nearest prime number greater than `dstParallelism * HashRangeMultipler`
- Uses error range for `idealSizePerTaskGroup` whose default value is set to 0.

### CODE REVIEW

File: runtime/common/src/main/java/edu/snu/coral/runtime/common/optimizer/pass/runtime/DataSkewRuntimePass.java [hunk 1]
```diff
@@ -116,12 +116,22 @@ public PhysicalPlan apply(final PhysicalPlan originalPlan, final Map<String, Lis
     // Do the optimization using the information derived above.
     final Long totalSize = aggregatedMetricData.stream().mapToLong(n -> n).sum(); // get total size
     final Long idealSizePerTaskGroup = totalSize / taskGroupListSize; // and derive the ideal size per task group
+    LOG.info("idealSizePerTaskgroup {} = {}(totalSize) / {}(taskGroupListSize)",
+        idealSizePerTaskGroup, totalSize, taskGroupListSize);
+
+    // Set an error rate for the ideal size calculated by math.
+    // Actual size we distribute per TaskGroup will set to range from lowerBoundSize to upperBoundSize.
+    final double errorRangeFactor = 0.0;
```
Comment (sanha) [inline runtime/common/src/main/java/edu/snu/coral/runtime/common/optimizer/pass/runtime/DataSkewRuntimePass.java]: Why don't we make this value as a configurable parameter?

**Comment by jeongyooneo** [comment]
@sanha Thanks! I've addressed the comment.

File: runtime/common/src/main/java/edu/snu/coral/runtime/common/optimizer/pass/runtime/DataSkewRuntimePass.java [hunk 2]
```diff
@@ -41,12 +43,30 @@
...
+  // The scope of actual size of data distributed to each TaskGroup is determined by this factor.
+  // lower bound: ideal size per TaskGroup - errorRangeFactor
+  // upper bound: ideal size per TaskGroup + errorRangeFactor
+  private final double errorRangeFactor;
 
   /**
    * Constructor.
    */
   public DataSkewRuntimePass() {
+    this.errorRangeFactor = 0.0;
```
Comment (sanha) [inline runtime/common/src/main/java/edu/snu/coral/runtime/common/optimizer/pass/runtime/DataSkewRuntimePass.java]: this(0.0) ?

File: runtime/common/src/main/java/edu/snu/coral/runtime/common/optimizer/pass/runtime/DataSkewRuntimePass.java [hunk 3]
```diff
@@ -41,12 +43,30 @@
...
+    this.eventHandlers = Stream.of(
+        DynamicOptimizationEventHandler.class
+    ).collect(Collectors.toSet());
+  }
+
+  /**
+   * Default constructor.
+   *
+   * @param errorRangeFactor    the factor that caps upper and lower bound of per TaskGroup-distributed data.
```
Comment (sanha) [inline runtime/common/src/main/java/edu/snu/coral/runtime/common/optimizer/pass/runtime/DataSkewRuntimePass.java]: Please check the indentation.

File: runtime/common/src/main/java/edu/snu/coral/runtime/common/optimizer/pass/runtime/DataSkewRuntimePass.java [hunk 4]
```diff
@@ -41,12 +43,30 @@
...
+  private final double errorRangeFactor;
 
   /**
    * Constructor.
    */
   public DataSkewRuntimePass() {
+    this.errorRangeFactor = 0.0;
+    this.eventHandlers = Stream.of(
+        DynamicOptimizationEventHandler.class
+    ).collect(Collectors.toSet());
```
Comment (wonook) [inline runtime/common/src/main/java/edu/snu/coral/runtime/common/optimizer/pass/runtime/DataSkewRuntimePass.java]: We can remove this redundant line if we just call `this(0.0)`.

File: compiler/optimizer/src/main/java/edu/snu/coral/compiler/optimizer/policy/DataSkewPolicy.java [hunk 5]
```diff
@@ -35,7 +35,7 @@
    */
   public DataSkewPolicy() {
     this.policy = new PolicyBuilder(true)
-        .registerRuntimePass(new DataSkewRuntimePass(), new DataSkewCompositePass())
+        .registerRuntimePass(new DataSkewRuntimePass(0.0), new DataSkewCompositePass())
```
Comment (wonook) [inline compiler/optimizer/src/main/java/edu/snu/coral/compiler/optimizer/policy/DataSkewPolicy.java]: I don't think we need to add the 0.0 here, as it's the default value.

File: runtime/common/src/main/java/edu/snu/coral/runtime/common/optimizer/pass/runtime/DataSkewRuntimePass.java [hunk 6]
```diff
@@ -130,15 +156,33 @@ public PhysicalPlan apply(final PhysicalPlan originalPlan, final Map<String, Lis
...
+        if (i == 1) {
+          finalSize = currentAccumulatedSize;
+        } else {
+          long currentSize = currentAccumulatedSize - sizePerTaskGroup.stream().mapToLong(l -> l).sum();
+          long oneStepBack = currentSize - aggregatedMetricData.get(finishingHashValue - 1);
+
+          // If the accumulated size for this TaskGroup exceeds upperBoundSize
+          // and taking off for one hash range doesn't violate lowerBoundSize, go one step back.
+          if (!(currentSize >= lowerBoundSize && currentSize <= upperBoundSize)) {
+            if (oneStepBack >= lowerBoundSize) {
```
Comment (wonook) [inline runtime/common/src/main/java/edu/snu/coral/runtime/common/optimizer/pass/runtime/DataSkewRuntimePass.java]: I think we can put this in another && operator.

File: runtime/common/src/main/java/edu/snu/coral/runtime/common/optimizer/pass/runtime/DataSkewRuntimePass.java (hunk 4 above)
Comment (wonook) [reply runtime/common/src/main/java/edu/snu/coral/runtime/common/optimizer/pass/runtime/DataSkewRuntimePass.java]: By the way, do you think 0.0 is a valid error range factor?

File: runtime/common/src/main/java/edu/snu/coral/runtime/common/optimizer/pass/runtime/DataSkewRuntimePass.java [hunk 7]
```diff
@@ -130,15 +156,33 @@ public PhysicalPlan apply(final PhysicalPlan originalPlan, final Map<String, Lis
...
+
+        long finalSize;
+        if (i == 1) {
+          finalSize = currentAccumulatedSize;
+        } else {
+          long currentSize = currentAccumulatedSize - sizePerTaskGroup.stream().mapToLong(l -> l).sum();
+          long oneStepBack = currentSize - aggregatedMetricData.get(finishingHashValue - 1);
+
+          // If the accumulated size for this TaskGroup exceeds upperBoundSize
+          // and taking off for one hash range doesn't violate lowerBoundSize, go one step back.
```
Comment (wonook) [inline runtime/common/src/main/java/edu/snu/coral/runtime/common/optimizer/pass/runtime/DataSkewRuntimePass.java]: Before, we compared the difference of currentAccumulatedSize and oneStepBack from the idealAccumulatedSize, to pick the one with the less difference. Now, we simply see if taking off one hash range doesn't violate the lowerBoundSize. Is this change better for evening out data?

File: compiler/optimizer/src/main/java/edu/snu/coral/compiler/optimizer/policy/DataSkewPolicy.java (hunk 5 above)
Comment (wonook) [reply compiler/optimizer/src/main/java/edu/snu/coral/compiler/optimizer/policy/DataSkewPolicy.java]: By the way, do you think 0.0 is a valid error range factor?

**Comment by jeongyooneo** [comment]
Thanks @wonook! As we've discussed offline, I've reverted to the good old style of calculating hash ranges. Distributing per-TaskGroup data evenly should be done elsewhere(for example at `Partitioner`), since `calculatingHashRange` simply divides at best effort given the already-partitioned hash ranges.

**Comment by wonook** [comment]
Looks good. I'll merge once the tests pass

**Pull request merged**